Query Cache

jamwt opened the pull request:
<!-- Describe your PR here. -->

Implementation of a query cache that can dramatically speed up re-rendering of values as SPA routes change and components are unmounted and remounted.

<!--
  The following applies to third-party contributors.
  Convex employees and contractors can delete or ignore.
-->

----

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

ianmacartney commented:
A couple thoughts just coming to me:
1. We could move this into the npm package, maybe namespaced under a `react/cache` entrypoint in a single file.
2. I wonder if we should change the tanstack-style `useQuery` hook in helpers to also do this, but only if it's in the cache provider - and maybe push folks to use that one. As it stands there will be 3 options for `useQuery`: vanilla, this one, and tanstack-style. there's a usecase for vanilla - if you don't want to use the caching - but that implies there should be an opt-out for the tanstack-style one too. So either 2 versions (TanStack to also get caching, vanilla if you want to opt out), or 4 versions (full permutations - or an extra arg to tanstack about opting out). I kinda wish I had gone with a single object arg for the tanstack one after all (`{ queryFn, queryArgs }`) 

Was it intentional to keep it out of the npm package for now, just to point people to it?

jamwt commented:
1. I just don't know how to do that, but otherwise -- yeah, would love to have it in the package.
2. Yeah, I debated the tanstack-style vs. this. Any ideas about how to support both without replicating a ton of code?

jamwt commented:
Okay, I

1. Refactored everything to be centered around `useQueries` so that the tanstack-style wrapper can use this cache.
2. Fixed error handling
3. Moved this into the `convex-helpers` package, in the react module.

ianmacartney commented on README.md:
```diff
@@ -113,6 +114,25 @@ Related files:
 - (optional)[useTypingIndicator.ts](./src/hooks/useTypingIndicator.ts) for specifically doing typing indicator presence.
 - (optional)[Facepile.tsx](./src/components/Facepile.tsx) for showing a facepile based on presence data. Intended to be used as an example to extend.
 
+## Query Caching
+
+A query cache implementation which persists subscriptions to the server for some expiration
+period even after app `useQuery` hooks have all unmounted. This allows very fast reloading
+of unevicted values during navigation changes, view changes, etc.
+
+Related files:
+
+- [provider.ts](./src/hooks/cache/provider.tsx) contains `ConvexQueryCacheProvider`,
+  a configurable cache provider you put in your react app's root.
+- [useQuery.tsx](./src/hooks/cache/useQuery.tsx) is a drop-in replacement for the
+  `convex/react` equivalent that utilizes this configured cache.
+- [CacheExample.tsx](./src/components/CacheExample.tsx) is a demo component demonstrating
+  the cache's `useQuery` to reuse app-side values when child components are
+  unmounted and re-mounted.
+
+To use the query cache, copy the entire `src/hooks/cache` directory into your project. The
+`cache/index.ts` re-exports the things you'll need.
+
```
let's move this to the README in packages/convex-helpers, and update the paths

jamwt replied:
Good point, will do.

ianmacartney commented on README.md:
```diff
@@ -29,6 +29,7 @@ find the npm package in [./packages/convex-helpers](./packages/convex-helpers).
 | [Presence](#presence)
 | [Throttling via single-flighting](#throttling-client-side-requests-by-single-flighting)
 | [Stable query results via useStableQuery](#stable-query-results-via-usestablequery)
+| [Query caching with ConvexQueryCacheProvider](#query-caching)
```
let's move this up to the list above, linking to the readme entry there

ianmacartney commented on packages/convex-helpers/package.json:
```diff
@@ -24,6 +24,18 @@
       "types": "./dist/react/sessions.d.ts",
       "default": "./dist/react/sessions.js"
     },
+    "./react/cache/core": {
+      "types": "./dist/react/cache/core.d.ts",
+      "default": "./dist/react/cache/core.js"
+    },
+    "./react/cache/hooks": {
+      "types": "./dist/react/cache/hooks.d.ts",
+      "default": "./dist/react/cache/hooks.js"
+    },
+    "./react/cache/provider": {
+      "types": "./dist/react/cache/provider.d.ts",
+      "default": "./dist/react/cache/provider.js"
+    },
```
Is it a React requirement that these be 3 separate files? I know the organization is nice, but I'd prefer to have it all in `react/cache.ts` so it's one entrypoint per feature

jamwt replied:
Yeah. SSR messes with this. "use client" (hook) code and provider code can't be in the same file.

ianmacartney replied:
... but the hooks file doesn't have "use client" at the top? At the least I'm taking out core, since none of the code is shared between them

ianmacartney commented on packages/convex-helpers/tsconfig.json:
```diff
@@ -4,16 +4,16 @@
     // "incremental": true,                              /* Save .tsbuildinfo files to allow for incremental compilation of projects. */
-    "composite": true,                                /* Enable constraints that allow a TypeScript project to be used with project references. */
+    "composite": true /* Enable constraints that allow a TypeScript project to be used with project references. */,
     // "tsBuildInfoFile": "./.tsbuildinfo",              /* Specify the path to .tsbuildinfo incremental compilation file. */
     // "disableSourceOfProjectReferenceRedirect": true,  /* Disable preferring source files instead of declaration files when referencing composite projects. */
     // "disableSolutionSearching": true,                 /* Opt a project out of multi-project reference checking when editing. */
     // "disableReferencedProjectLoad": true,             /* Reduce the number of projects loaded automatically by TypeScript. */
 
     /* Language and Environment */
-    "target": "ESNext",                                  /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */
+    "target": "ESNext" /* Set the JavaScript language version for emitted JavaScript and include compatible library declarations. */,
```
I think these are unintentional - maybe our prettier config now fights with the default tsconfig formatting?

jamwt replied:
I dunno. I can revert... is that what you're suggesting?
